pkg/util/log_support.go

```diff
@@ -0,0 +1,347 @@
+	// be desired.
+	DisableSorting bool
+
+	// QuoteEmptyFields will wrap empty fields in quotes if true
+	QuoteEmptyFields bool
+
+	// Whether the logger's out is to a terminal
+	isTerminal bool
+
+	sync.Once
+}
+
+// LogSetUp sets up UserOut and log loggers as needed by ddev
+func LogSetUp() {
+	// Use stdout instead of stderr for all user output
+	log.SetOutput(os.Stdout)
+	UserOut.Out = os.Stdout
+
+	if !JSONOutput {
+		UserOut.Formatter = UserOutFormatter
+	} else {
+		UserOut.Formatter = &JSONFormatter{}
+	}
+
+	UserOutFormatter.DisableTimestamp = true
+	// Always use log.DebugLevel for UserOut
+	UserOut.Level = log.DebugLevel // UserOut will by default always output
+
+	// But we use custom DRUD_DEBUG-settable loglevel for log
+	logLevel := log.InfoLevel
+	drudDebug := os.Getenv("DRUD_DEBUG")
+	if drudDebug != "" {
+		logLevel = log.DebugLevel
+	}
+	log.SetLevel(logLevel)
+}
+
+func (f *TextFormatter) init(entry *log.Entry) {
+	if entry.Logger != nil {
+		f.isTerminal = f.checkIfTerminal(entry.Logger.Out)
+	}
+}
+
+func (f *TextFormatter) checkIfTerminal(w io.Writer) bool {
+	switch v := w.(type) {
+	case *os.File:
+		return terminal.IsTerminal(int(v.Fd()))
+	default:
+		return false
+	}
+}
+
+// Format renders a single log entry. A key named "raw" is discarded here as it's for json consumption.
+func (f *TextFormatter) Format(entry *log.Entry) ([]byte, error) {
+	var b *bytes.Buffer
+	keys := make([]string, 0, len(entry.Data))
+	for k := range entry.Data {
+		// Discard key named "raw" as it's for json consumption only.
+		if k != "raw" {
+			keys = append(keys, k)
+		}
+	}
+
+	if !f.DisableSorting {
+		sort.Strings(keys)
+	}
+	if entry.Buffer != nil {
+		b = entry.Buffer
+	} else {
+		b = &bytes.Buffer{}
+	}
+
+	prefixFieldClashes(entry.Data)
+
+	f.Do(func() { f.init(entry) })
+
+	isColored := (f.ForceColors || f.isTerminal) && !f.DisableColors
+
+	timestampFormat := f.TimestampFormat
+	if timestampFormat == "" {
+		timestampFormat = defaultTimestampFormat
+	}
+	if isColored {
+		f.printColored(b, entry, keys, timestampFormat)
+	} else {
+		if !f.DisableTimestamp {
+			f.appendKeyValue(b, "time", entry.Time.Format(timestampFormat))
+		}
+		//f.appendKeyValue(b, "level", entry.Level.String())
+		if entry.Message != "" {
+			//f.appendKeyValue(b, "msg", entry.Message)
+			f.appendValue(b, entry.Message)
+		}
+		for _, key := range keys {
+			f.appendKeyValue(b, key, entry.Data[key])
+		}
+	}
+
+	b.WriteByte('\n')
+	return b.Bytes(), nil
+}
+
+func (f *TextFormatter) printColored(b *bytes.Buffer, entry *log.Entry, keys []string, timestampFormat string) {
+	var levelColor int
+	switch entry.Level {
+	case log.DebugLevel:
+		levelColor = gray
+	case log.WarnLevel:
+		levelColor = yellow
+	case log.ErrorLevel, log.FatalLevel, log.PanicLevel:
+		levelColor = red
+	default:
+		levelColor = nocolor
+	}
+
+	fmt.Fprintf(b, "\x1b[%dm%s\x1b[0m ", levelColor, entry.Message)
```

Comment:
i'm kind of surprised we have to re-implement handling of so many things like colored output or quoting here. is this also adapted from logrus code?

Reply:
Logrus would have been emitting and coloring the log level like [FATA] would be in red, we're coloring the whole thing. That's the change from logrus.

Reply:
It'd be helpful to have explanations like this in code comments. I still don't have a firm grasp of what's copied as-is from logrus vs what's been modified, and it'd help to be able to make this determination quickly months later when we've forgotten this discussion.

Reply:
I reorganized the formatters (while putting in a separate package for @beeradb ) so that the names are the same as the original files, and there are checked-in versions of the original files in the sirupsen_orig directory. That makes it easy to diff and see what has changed.